### ISSUE TITLE
Same MVT system, different behaviour on 3.x and 4.x

### ISSUE THREAD

**Issue by cloudynetwork** [SDL-Hercules-390/hyperion#174]
I'm sorry about the vagueness of this but any guidance on the dumps etc I'd need to provide to help resolve would be great. 

I have an MVT system with HASP that runs mostly fine on 3.x...however, if I run the exact same system on 4.x I get a lot of HASP related ABENDs / Core Dump attempts performing exactly the same activities. This is fairly repeatable. 

What information can I provide if this is of interest as a possible issue?

**Comment by bluecmd** [comment]
I'm not an author or anything, but I can try to assist a bit in the initial triage. You'll most likely be required to submit exactly what versions you are using, full logs from Hercules on both versions, and the full configuration files used. As well as exact logs of the issues you mention.

Note that the configuration file format has changed between the 3.x and 4.x series so if you haven't looked at the log to determine if Hercules/Hyperion thinks that some values are now deprecated that might be a good start.

**Comment by Fish-Git** [comment]
Closely reviewing the **[Release Notes](https://sdl-hercules-390.github.io/html/hercrnot.html)** might be helpful as well. @bluecmd is correct: much has changed between 3.x and 4.x. You will very likely need to make adjustments to your configuration file as a result.

**Comment by cloudynetwork** [comment]
Understood. I am in fact using the same config file so will take the
suggested steps and retest.

**Comment by Fish-Git** [comment]
___
_**I am re-opening this issue as, clearly, there is still an unresolved Hercules problem (bug) that still needs to be found and fixed.**_
___
There appears to still be an issue (bug) in SDL Hercules 4.x Hyperion _(or a new bug was introduced)_ that is negatively impacting MVT. Refer the the recent Hercules-390 group thread: "[MVT with HASP on SDL Hercules 4.4 and above](https://hercules-390.groups.io/g/group/message/5446)" for details.

Briefly, SDL Hercules 4.3 Hyperion works perfect. Beginning with SDL Hercules 4.4 Hyperion however, it seems a bug was introduced that is now causing MVT to abend under certain conditions. In the simplest case which, with the help of René Ferland, I was able to easily and consistently reproduce the problem:

```
How to reproduce the bug
========================

1. Start Hercules on the configuration file

2. Start a 3270 connect on device 010

3. ipl 150

4. At message IEA101A, just type Enter

5. At message IEE114A, reply
 
   r 00,auto=n

6. Press PF12 when you see message IEE159E in lower right corner

7. At message SPECIFY HASP OPTIONS, reply

   r 01,cold,format

8. When you see message ENTER HASP REQUESTS, reply

   $s

   to start HASP.

9. When you see message ALL AVAILABLE FUNCTIONS COMPLETE,
   you can submit the job from the card reader

   devinit 00c jcl\COB-CARS.txt eof

Rene FERLAND, Montreal
```

_(Fish note: you can also submit the card reader job via HercRdr or any other tool that allows you to submits jobs through Hercules's sockdev hot reader.)_

Here are the files needed to reproduce the problem:

* [MVT#1.zip](https://github.com/SDL-Hercules-390/hyperion/files/10810147/MVT.1.zip)
* [MVT#2.zip](https://github.com/SDL-Hercules-390/hyperion/files/10810151/MVT.2.zip)

_(The .zip file that Rene sent me was too large to upload to GitHub, so I created two separate .zip files instead, each one with a different set of the files that are needed to reproduce the problem.)_

Here is a partial Hercules log of the problem occurring:

<pre>
/ $ 1.12.49 INIT 1  IDLE-CLASS=RA9
/ $ 1.12.49 INIT 2  IDLE-CLASS=RA9
/ $ 1.12.49 INIT 3  IDLE-CLASS=RA9
/ $ 1.12.49 PRINTER1 IDLE - CLASS=AJ
/ $ 1.12.49 PRINTER2 IDLE - CLASS=AJ
/ $ 1.12.49 ALL AVAILABLE FUNCTIONS COMPLETE
HHC01040I 0:000C COMM: client Win7, IP 127.0.0.1 connected to device 3505
/ $ 1.13.01 JOB    1 ON READER1  -- R13264J   R FERLAND
HHC01206I 0:000C Card: client Win7, IP 127.0.0.1 disconnected from device 3505
/ $ 1.13.01 JOB    1 -- R13264J  -- BEGINNING EXEC - INIT 1 - CLASS A
<b>HHC00801I Processor CP00: Operation exception interruption code 0001 ilc 2
HHC02269I GR00=00032888 GR01=007BB0A0 GR02=007BB670 GR03=007BB098
HHC02269I GR04=00000000 GR05=00000050 GR06=00000050 GR07=00000028
HHC02269I GR08=127FE3B0 GR09=007BB6F0 GR10=807ADC60 GR11=407AFBAA
HHC02269I GR12=007B9800 GR13=007BB2B8 GR14=507FE3E2 GR15=007C0298
HHC02271I CR00=800000E0 CR01=00000000 CR02=FFFFFFFF CR03=00000000
HHC02271I CR04=00000000 CR05=00000000 CR06=00000000 CR07=00000000
HHC02271I CR08=00004000 CR09=00000000 CR10=00000000 CR11=00000000
HHC02271I CR12=00000000 CR13=00000000 CR14=CFC00000 CR15=0002AB28
HHC02324I PSW=FF05001F807FDB52 INST=0008         ????? ,                      ?</b>
/   1.13.01 IEF450I HOSRDR  .700     .         ABEND S0C1       TIME=01.13.01
</pre>

Based on analysis by [Kevin Leonard](https://hercules-390.groups.io/g/group/message/5463), it appears the bug, whatever it is, is causing MVT to erroneously branch into the middle of an instruction. Based on that, the bug appears to be either some type of incorrect ILC _(Instruction Length Code)_ handling issue, or a PSW handling issue, thereby causing a bad instruction address to be stored in the PSW _(which MVT then tries using,  causing the crash)_.

_**I should also mention** (in case it might be related to the problem)_ that MVT runs in `BC` mode, not `EC` mode _(as should be obvious based on the PSW)._

_(**Note:** The `TRTR` issue that Kevin raises in his post is a [Red Herring](https://en.wikipedia.org/wiki/Red_herring), a distraction that has **absolutely nothing** to do with the problem. It is simply a [minor cosmetic issue](https://hercules-390.groups.io/g/group/message/5466).)_

I am currently actively trying to debug this issue, but thought I'd post it officially here _(since it is after all a bona fide unresolved Hercules bug)_ in the hope that someone _(or several someones!)_ might help me figure out what the FRICK we _(probably me!)_ did to cause this problem to appear.

Thanks.

My apologies to @cloudynetwork _(Peter Joseph)_ for prematurely closing this GitHub Issue of his, and for incorrectly classifying it as a User Error _(which it most definitely is **NOT!**)._

**Comment by Fish-Git** [comment]
_**FYI:**_ When it works (e.g. when SDL Hercules **4.3** Hyperion is used), the Hercules log looks like this:

```
/ $ 0.55.05 INIT 1  IDLE-CLASS=RA9
/ $ 0.55.05 INIT 2  IDLE-CLASS=RA9
/ $ 0.55.05 INIT 3  IDLE-CLASS=RA9
/ $ 0.55.05 PRINTER1 IDLE - CLASS=AJ
/ $ 0.55.05 PRINTER2 IDLE - CLASS=AJ
/ $ 0.55.05 ALL AVAILABLE FUNCTIONS COMPLETE
HHC01040I 0:000C COMM: client Win7, IP 127.0.0.1 connected to device 3505
HHC01206I 0:000C Card: client Win7, IP 127.0.0.1 disconnected from device 3505
/ $ 0.55.11 JOB    1 ON READER1  -- R13264J   R FERLAND
/ $ 0.55.11 JOB    1 -- R13264J  -- BEGINNING EXEC - INIT 1 - CLASS A
/   0.55.12 JOB    1 STEP: R13264J  CARS     COB      ET=00.00.00 RC=0000
/   0.55.12 JOB    1 STEP: R13264J  CARS     GO       ET=00.00.00 RC=0000
/ $ 0.55.12 INIT 1  IDLE-CLASS=RA9
/ $ 0.55.12 JOB    1 END EXECUTION.
/ $ 0.55.12 JOB    1 PRINTING ON PRINTER1
/ $ 0.55.12 PRINTER1 IDLE - CLASS=AJ
/ $ 0.55.12 JOB    1 IS PURGED
```

We can see HercRdr connecting to the hot reader on device 0:000C _(message `HHC01040I`)_, and then immediately disconnecting after submitting the job _(message `HHC01206I`)_. The job then runs to successful completion. That's how things _should_ work.

When SDL Hercules **4.4** Hyperion or greater is used however, we get the S0C1 abend. `:(`

**Comment by Fish-Git** [comment]
Fixed by commit 2d443708f596dec8953758cabb4999c840f0d093.

Closing.
